### PR TITLE
Implement lifted_struct_loss

### PR DESCRIPTION
**Pull request by tabshaikh**
Related Issue Number: #26

### CODE REVIEW

**Comment by tabshaikh** [comment]
CLA signed

**Comment by facaiy** [comment]
Hi, Tabish. Can you merge the latest master branch, and add unit tests necessary? Thank you for your contribution :-)

**Comment by tabshaikh** [comment]
Yes I will finish this till tomorrow

**Comment by tabshaikh** [comment]
@facaiy @seanpmorgan  the pr is ready for review, do I have to change the BUILD file also.

**Comment by seanpmorgan** [comment]
Short answer is yes, the BUILD file has to be adjusted so we can run the test cases using bazel. I plan to push a some contribution guidelines tomorrow so that should make it more clear.

File: tensorflow_addons/losses/python/metric_learning.py [hunk 1]
```diff
@@ -72,3 +74,30 @@ def pairwise_distance(feature, squared=False):
     pairwise_distances = math_ops.multiply(pairwise_distances,
                                            mask_offdiagonals)
     return pairwise_distances
+
+@tf.function
+def pairwise_distance_np(feature, squared=False):
```
Comment (facaiy) [inline tensorflow_addons/losses/python/metric_learning.py]: Don't move it here. I prefer to copy `pairwise_distance_np` function in your lifted_test.py. In the future, we should remove it and modify our test cases to check explicit results directly.

File: tensorflow_addons/losses/python/metric_learning.py [hunk 2]
```diff
@@ -72,3 +74,30 @@ def pairwise_distance(feature, squared=False):
     pairwise_distances = math_ops.multiply(pairwise_distances,
                                            mask_offdiagonals)
     return pairwise_distances
+
+@tf.function
```
Comment (facaiy) [inline tensorflow_addons/losses/python/metric_learning.py]: remove the decorator

File: tensorflow_addons/losses/python/lifted.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    Returns:
+        lifted_loss: tf.float32 scalar.
+    """
+    
+    def __init__(self, margin=1.0, name=None):
+        super(LiftedStructLoss, self).__init__(
+            lifted_struct_loss,
+            name=name,
+            reduction=losses_impl.ReductionV2.NONE,
+            margin=margin)
```
Comment (facaiy) [inline tensorflow_addons/losses/python/lifted.py]: add newline in the end

File: tensorflow_addons/losses/python/lifted.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  return lifted_loss
+
+@keras_utils.register_keras_custom_object
+class LiftedStructLoss(losses.LossFunctionWrapper):
+    """Computes the lifted structured loss.
+    The loss encourages the positive distances (between a pair of embeddings
+    with the same labels) to be smaller than any negative distances (between a
+    pair of embeddings with different labels) in the mini-batch in a way
+    that is differentiable with respect to the embedding vectors.
+    See: https://arxiv.org/abs/1511.06452.
```
Comment (facaiy) [inline tensorflow_addons/losses/python/lifted.py]: Add blank line to separate those block

File: tensorflow_addons/losses/python/metric_learning.py (hunk 2 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/python/metric_learning.py]: @facaiy removed the function only

File: tensorflow_addons/losses/python/metric_learning.py (hunk 1 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/python/metric_learning.py]: Okay done

File: tensorflow_addons/losses/python/lifted.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import tensorflow as tf
+from tensorflow.python.framework import dtypes
+from tensorflow.python.keras import losses
+from tensorflow.python.keras.utils import generic_utils
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops.losses import losses_impl
+from tensorflow_addons.losses.python import metric_learning
+from tensorflow_addons.utils.python import keras_utils
+
```
Comment (facaiy) [inline tensorflow_addons/losses/python/lifted.py]: Please add one more blank line here, and fix all. https://www.python.org/dev/peps/pep-0008/

> Surround top-level function and class definitions with two blank lines.

File: tensorflow_addons/losses/python/lifted.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+@keras_utils.register_keras_custom_object
+@tf.function
+def lifted_struct_loss(labels, embeddings, margin=1.0):
+  """Computes the lifted structured loss.
+  Args:
```
Comment (facaiy) [inline tensorflow_addons/losses/python/lifted.py]: Please fix all docstring: use blank line to separate them. `pairwise_distance_np` is a good example

File: tensorflow_addons/losses/python/lifted.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        lifted_loss: tf.float32 scalar.
+    """
+    
+    def __init__(self, margin=1.0, name=None):
+        super(LiftedStructLoss, self).__init__(
+            lifted_struct_loss,
+            name=name,
+            reduction=losses_impl.ReductionV2.NONE,
+            margin=margin)
+
```
Comment (facaiy) [inline tensorflow_addons/losses/python/lifted.py]: Sorry for misunderstanding. Add a newline character ('\n'), instead of a new line. See https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

File: tensorflow_addons/losses/python/lifted.py (hunk 6 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/python/lifted.py]: Okay thanks :)

File: tensorflow_addons/losses/python/lifted.py (hunk 5 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/python/lifted.py]: Done

File: tensorflow_addons/losses/python/lifted.py (hunk 7 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/python/lifted.py]: There was bug in vscode which would not add the newline character at the end of the file https://github.com/Microsoft/vscode/issues/58276 so therefore it got me confused too :)

Comment (tabshaikh) [reply tensorflow_addons/losses/python/lifted.py]: But this is done

**Comment by tabshaikh** [comment]
@facaiy I have made the changes and also added the build file

File: tensorflow_addons/losses/BUILD [hunk 8]
```diff
@@ -28,3 +28,30 @@ py_test(
     ],
     srcs_version = "PY2AND3",
 )
+
+py_library(
+    name = "lifted_py",
```
Comment (facaiy) [inline tensorflow_addons/losses/BUILD]: Please put lifted.py in the first py_library `"losses_py"`, and delete the py_library. We can split them in the future if necessary.

File: tensorflow_addons/losses/BUILD [hunk 9]
```diff
@@ -28,3 +28,30 @@ py_test(
...
+py_test(
+    name = "lifted_py_test",
+    size = "small",
+    srcs = [
+        "python/lifted_test.py",
+    ],
+    main = "python/lifted_test.py",
+    deps = [
+        ":losses_py",
+    ],
```
Comment (facaiy) [inline tensorflow_addons/losses/BUILD]: The build config seems not totally correct. Please use bazel to run tests at first and make sure all pass, see more details at: 
+ https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md
+ https://docs.bazel.build/versions/master/bazel-overview.html

File: tensorflow_addons/losses/python/lifted.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    The loss encourages the positive distances (between a pair of embeddings
+    with the same labels) to be smaller than any negative distances (between a
+    pair of embeddings with different labels) in the mini-batch in a way
+    that is differentiable with respect to the embedding vectors.
+    See: https://arxiv.org/abs/1511.06452.
+
+    Args:
+        labels: 1-D tf.int32 `Tensor` with shape [batch_size] of
+        multiclass integer labels.
```
Comment (facaiy) [inline tensorflow_addons/losses/python/lifted.py]: Please make sure all doc format are consistent, refer to https://github.com/tensorflow/addons/pull/28/files

File: tensorflow_addons/losses/python/lifted_test.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+                labels=convert_to_tensor(labels),
+                embeddings=convert_to_tensor(embedding),
+                margin=margin)
+            loss_tf = loss_tf.eval()
+            self.assertAllClose(loss_np, loss_tf)
+
+
+if __name__ == '__main__':
+    test.main()
+
```
Comment (facaiy) [inline tensorflow_addons/losses/python/lifted_test.py]: delete the last blank line

File: tensorflow_addons/losses/BUILD (hunk 8 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/BUILD]: Done

File: tensorflow_addons/losses/BUILD (hunk 9 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/BUILD]: You were right i was building the test incorrectly, thanks @facaiy

File: tensorflow_addons/losses/python/lifted_test.py (hunk 11 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/python/lifted_test.py]: done

File: tensorflow_addons/losses/python/lifted.py (hunk 10 above)
Comment (tabshaikh) [reply tensorflow_addons/losses/python/lifted.py]: Do you want it in a single line

**Comment by tabshaikh** [comment]
@facaiy Build now passing 
test_passed

**Comment by facaiy** [comment]
Please wait a moment, I'll merge it later after I reset CLA status.

**Comment by facaiy** [comment]
@karmel @av8ramit Hi, Karmel, Amit. Could you help reset CLA status?

**Comment by tabshaikh** [comment]
Thank you so much @seanpmorgan @facaiy @karmel for merging this, finally one pr merged. Yay!!!!!! :)

**Comment by sarim-zafar** [comment]
@tabshaikh I recently tried using lifted struct loss and as far as I understand it can be replaced with the TripletSemiHardLoss as is without any code modification. But when I try to compile the model it gives me a shape error. I am using tf 2.0 on collab. Can you please provide some sample code perhaps?
